src/IO/readFloatText.h

```diff
@@ -148,11 +148,11 @@ ReturnType readFloatTextPreciseImpl(T & x, ReadBuffer & buf)
     static_assert('a' > '.' && 'A' > '.' && '\n' < '.' && '\t' < '.' && '\'' < '.' && '"' < '.', "Layout of char is not like ASCII");
 
     static constexpr bool throw_exception = std::is_same_v<ReturnType, void>;
-
-    /// Fast path (avoid copying) if the buffer have at least MAX_LENGTH bytes.
     static constexpr int MAX_LENGTH = 316;
+    ReadBufferFromMemory * buf_from_memory = dynamic_cast<ReadBufferFromMemory *>(&buf);
 
-    if (likely(!buf.eof() && buf.position() + MAX_LENGTH <= buf.buffer().end()))
+    /// Fast path (avoid copying) if the buffer have at least MAX_LENGTH bytes or buf is ReadBufferFromMemory
+    if (likely(!buf.eof() && (buf_from_memory || buf.position() + MAX_LENGTH <= buf.buffer().end())))
```

Comment:
I don't think it's the right way to fix it. If buf is not `ReadBufferFromMemory`, we will still have this bug. The actual problem is in slow path here:
https://github.com/ClickHouse/ClickHouse/blob/5923e1b1167b1bf755654e3dcb46c13106ec164b/src/IO/readFloatText.h#L241-L256

We don't check that data data copied to `tmp_buf` was used during float parsing. We can change ckech on line 250:
```
        if (unlikely(res.ec != std::errc()) || res.ptr - tmp_buf != num_copied_chars)
            ...
```

Reply:
@Avogar If `readFloatTextPrecise` guarantees that `ReadBuffer & in` is all read after parsing, then you solution is right. Otherwise, we should use PeekableReadBuffer to mark the position before parsing, which is too heavy for current function. So I fix it using `ReadBufferFromMemory`(notice that `readFloatTextPreciseImpl` is only used in ConvertThroughParsing::execute, and all ReadBuffer passed to `readFloatTextPreciseImpl` is `ReadBufferFromMemory`)

Reply:
> Otherwise, we should use PeekableReadBuffer to mark the position where the actual float parsing is finished, and rollback to it if not all chars copied to buffer are consumed

Ideally yes, we should rollback to the position when actual float parsing is finished, but right now we can read from the buffer more data than was used for float parsing and this extra data will be just silently ignored, like in example `2020-01-01` part `-01-01` will be read and just ignored, so it  won't be available after  `readFloatTextPreciseImpl` is finished. So I think we better throw an exception in this case instead of silently consuming extra data

Reply:
But current checking of whether there are extra data after parsing is in `ConvertThroughParsing::isAllRead`.  If we throw exception in `readFloatTextPreciseImpl`, it will break the rules.

Reply:
> it will break the rules

What rules? `readFloatTextPreciseImpl ` should parse correct float value from read data or throw an exception, it should not silently read some extra data without an exception.
The only think that will change is where we throw and exception and an exception message. And we can just improve exception message in `readFloatTextPreciseImpl` (make it like: `Cannot parse flloating point value here: <data_copied_to_tmp_buffer>`).

Reply:
But we don't throw an exception in `ConvertThroughParsing::isAllRead` if `readFloatTextPreciseImpl` read extra bytes in tmp buffer in slow path. If we have `2020-01-01` and execute slow path in `readFloatTextPreciseImpl` we still get no excaption and silently ignored data

Reply:
In slow path there is no way we can rely on `ConvertThroughParsing::isAllRead` because we can read extra data from the buffer that won't be used in float parsing. And `readFloatTextPreciseImpl ` should be responsible to throw an exception in this case, because we can't do it in any other place

Reply:
And we should think about `readFloatTextPreciseImpl` in isolation as a function that parses float from the data and throw an exception if parsing goes wrong. Reading extra bytes from buffer and ignoring them is something wrong and we should throw an exception

Reply:
So, I suggest keep check for `ReadBufferFromMemory` with a comment that this is an optimization to avoid copying data, but also add a check with an exception for slow path

Reply:
Get it! I'll update codes and pls review later.